Add Tutorials section to doc website (#4934)
openshift/odo: website/docs/tutorials/deploying-java-app-with-database.md

---
title: Deploying a Java OpenLiberty application with PostgreSQL
sidebar_position: 1
---

This tutorial illustrates deploying a [Java OpenLiberty](https://openliberty.io/) application with odo and linking it to an in-cluster PostgreSQL service in the minikube environment.

There are two roles in this example:

1. Cluster Admin - Prepare the cluster by installing the required operators on the cluster.
2. Application Developer - Imports a Java application, creates a Database instance, and connect the application to the Database instance.

## Cluster admin
---

> This section assumes that you have installed [minikube and configured it](../getting-started/cluster-setup/kubernetes.md).

Reviewer: there is no content in the page this links to

Reviewer 2: It'll be added as a part of https://github.com/openshift/odo/pull/4973. So we are good to have it, IMO.

[//]: # (Move this section to Architecture > Service Binding or create a new Operators doc)

We will be using Operators in this guide. An Operator helps in deploying the instances of a given service, for example PostgreSQL, MySQL, Redis.

Furthermore, these Operators are "bind-able". Meaning, if they expose information necessary to connect to them, odo can help connect your component to their instances.

[//]: # (Move until here)

> See the [Operator installation guide](../getting-started/cluster-setup/kubernetes.md) to install and configure an Operator in the Kubernetes cluster.

Reviewer: another missing page

Reviewer 2: https://github.com/openshift/odo/pull/4934#discussion_r686666032

    [ERROR] testGetAllPeople  Time elapsed: 0.33 s  <<< FAILURE!
    org.opentest4j.AssertionFailedError: Expected at least 2 people to be registered, but there were only: [] ==> expected: <true> but was: <false>
    at org.example.app.it.DatabaseIT.testGetAllPeople(DatabaseIT.java:57)
    
    [ERROR] testGetPerson  Time elapsed: 0.047 s  <<< ERROR!
    java.lang.NullPointerException
    at org.example.app.it.DatabaseIT.testGetPerson(DatabaseIT.java:41)
    
    [INFO]
    [INFO] Results:
    [INFO]
    [ERROR] Failures:
    [ERROR]   DatabaseIT.testGetAllPeople:57 Expected at least 2 people to be registered, but there were only: [] ==> expected: <true> but was: <false>
    [ERROR] Errors:
    [ERROR]   DatabaseIT.testGetPerson:41 NullPointer
    [INFO]
    [ERROR] Tests run: 2, Failures: 1, Errors: 1, Skipped: 0
    [INFO]
    [ERROR] Integration tests failed: There are test failures.
    ```
   Note: This error will be fixed at a later stage in the tutorial when we connect a database instance to this application.

6. You can also create a URL with `odo` to access the application:
    ```shell

>     odo url create --host $(minikube ip).nip.io

Author: @dharmit We will need to enable ingress and ingress-dns addons on minikube for this to work, won't we?

Reviewer 2: Yes. But that should be covered in the Kubernetes cluster setup guide in Getting Started section.

---
title: Deploying a Java OpenLiberty application with PostgreSQL
sidebar_position: 1
---

This scenario illustrates deploying a Java application with odo and linking it to an in-cluster PostgreSQL service in the minikube environment.

In this example there are two roles:

1. Cluster Admin - Prepare the cluster by installing the required operators into the cluster
2. Application Developer - Imports a Java application, creates a Database instance, and connect the application to the Database instance.

## Cluster admin
---

> This section assumes that you have installed minikube and configured it. See [Getting Started > Cluster Setup > Kubernetes](../getting-started/cluster-setup/kubernetes.md).

Reviewer 2: ```suggestion
This section assumes that you have installed minikube and configured it. See the section on setting up[Kubernetes](../getting-started/cluster-setup/kubernetes.md) if you have not already configured it.
```

The cluster admin must install two Operators into the cluster:

> 1. Operator Backed Service

Reviewer 2: No. Cluster admin needs to install the specific Postgres Operator that this example is using.

> 2. Service Binding Operator

Reviewer 2: This should be a link to the installation pages covering SBO installation on minikube and CRC.

Author: I added a separate note at the end of this section referring to the operator installation.

> An _Operator Backed Service_ is an operator that helps in deploying instances of a given service, for example PostgreSQL, MySQL, Redis.

Reviewer 2: This statement needs to change once the point 1 above changes. It will explain why the Postgres Operator is required.

> See the [Operator installation guide](../getting-started/cluster-setup/kubernetes.md) to install and configure an Operator in the Kubernetes cluster, if you have not already done so.

Reviewer 2: This part should go into guide on Operators in Architecture section. Can you add a TODO here so that we address it at some point?

Furthermore, these operators are "bind-able" as they expose information necessary for an application to connect to their instances.

> We will use Dev4Devs PostgreSQL Operator found in the [OperatorHub](https://operatorhub.io) to demonstrate a sample use case.

Reviewer 2: ```suggestion
We will use [Dev4Devs PostgreSQL Operator](https://operatorhub.io/operator/postgresql-operator-dev4devs-com) found on the [OperatorHub](https://operatorhub.io) to demonstrate a sample use case.
```

> Service Binding Operator is an operator that helps in easily binding an application to other Operator Backed Services. It accomplishes this through automatically collecting binding information and sharing with an application to bind it with operator managed backing services

Reviewer 2: This becomes redundant when we explain how to install SBO and why in the installation pages for miniube & CRC.

### Installing the Operator Backed Service

* Run the following `kubectl` command to make the PostgreSQL Operator available in `my-postgresql-operator-dev4devs-com` namespace of your minikube cluster:
```shell

>   $ kubectl create -f https://operatorhub.io/install/postgresql-operator-dev4devs-com.yaml

Reviewer 2: So that user can directly copy the code
```suggestion
kubectl create -f https://operatorhub.io/install/postgresql-operator-dev4devs-com.yaml
```

  ```

**Note**: The `my-postgresql-operator-dev4devs-com` Operator will be installed in the `my-postgresql-operator-dev4devs-com` namespace and will be usable from this namespace only.

> ### Installing the Service Binding Operator

Author: I wonder if we will need a separate section on installing SBO and PostgreSQLO if we already point to the installation guides in the section above.

Reviewer 2: No we don't. We should add a link to SBO installation that we plan to cover in kubernetes and openshift installation guides in "Getting Started" section.

* Run the following `kubectl` command to make the Service Binding Operator available in all namespaces on your minikube:
    ```shell

>     $ kubectl create -f https://operatorhub.io/install/service-binding-operator.yaml

Reviewer 2: So that user can directly copy the code
```suggestion
kubectl create -f https://operatorhub.io/install/service-binding-operator.yaml
```

    ```

>   Refer to [Getting Started > Cluster Setup > Operators](../getting-started/cluster-setup/operators.md) for more information on installing operators.

Reviewer 2: ```suggestion
  Refer to [Operators guide](../getting-started/cluster-setup/operators.md) for more information on installing operators.
```

## Application Developer
---

> This section assumes that you have installed `odo`. See [Getting Started > Installation](../getting-started/installation.md).

Reviewer 2: ```suggestion
This section assumes that you have installed `odo`. See the [Installation guide](../getting-started/installation.md) if you haven't already installed odo.
```

Since the PostgreSQL Operator installed in above step is available only in `my-postgresql-operator-dev4devs-com` namespace, ensure that `odo` uses this namespace to perform any tasks:

```shell

> $ odo project set my-postgresql-operator-dev4devs-com

Reviewer 2: ```suggestion
odo project set my-postgresql-operator-dev4devs-com
```

```
## Importing the demo Java MicroService JPA application

In this example we will use odo to manage a sample [Java MicroServices JPA application](https://github.com/OpenLiberty/application-stack-samples.git).

1. Clone the sample application to your system:
    ```shell

>     $ git clone https://github.com/OpenLiberty/application-stack-samples.git

Reviewer 2: ```suggestion
git clone https://github.com/OpenLiberty/application-stack-samples.git
```

    ```

2. Go to the sample JPA app directory:
    ```shell

>     $ cd ./application-stack-samples/jpa

Reviewer 2: ```suggestion
cd ./application-stack-samples/jpa
```

    ```

3. Initialize the project:
    ```shell

>     $ odo create java-openliberty mysboproj

Reviewer 2: ```suggestion
odo create java-openliberty mysboproj
```

    ```

4. Push the application to the cluster:
    ```shell

>     $ odo push

Reviewer 2: ```suggestion
odo push
```

    ```

5. The application is now deployed to the cluster - you can view the status of the cluster and the application test results by streaming the OpenShift logs to the terminal.
    ```shell

>     $ odo log

Reviewer 2: ```suggestion
odo log
```

    [INFO] [err]    ... 86 more
    [ERROR] Tests run: 2, Failures: 1, Errors: 1, Skipped: 0, Time elapsed: 0.706 s <<< FAILURE! - in org.example.app.it.DatabaseIT
    [ERROR] testGetAllPeople  Time elapsed: 0.33 s  <<< FAILURE!
    org.opentest4j.AssertionFailedError: Expected at least 2 people to be registered, but there were only: [] ==> expected: <true> but was: <false>
    at org.example.app.it.DatabaseIT.testGetAllPeople(DatabaseIT.java:57)
    
    [ERROR] testGetPerson  Time elapsed: 0.047 s  <<< ERROR!
    java.lang.NullPointerException
    at org.example.app.it.DatabaseIT.testGetPerson(DatabaseIT.java:41)
    
    [INFO]
    [INFO] Results:
    [INFO]
    [ERROR] Failures:
    [ERROR]   DatabaseIT.testGetAllPeople:57 Expected at least 2 people to be registered, but there were only: [] ==> expected: <true> but was: <false>
    [ERROR] Errors:
    [ERROR]   DatabaseIT.testGetPerson:41 NullPointer
    [INFO]
    [ERROR] Tests run: 2, Failures: 1, Errors: 1, Skipped: 0
    [INFO]
    [ERROR] Integration tests failed: There are test failures.
    ```

6. You can also create an ingress URL with `odo` to access the application:
    ```shell

>     $ odo url create --host $(minikube ip).nip.io

Reviewer 2: ```suggestion
odo url create --host $(minikube ip).nip.io
```

    ```

7. Push the URL to activate it:
    ```shell

>     $ odo push

Reviewer 2: ```suggestion
odo push
```

    ```

8. Display the created URL:
    ```shell
    $ odo url list
    ```

    You will see a fully formed URL that can be used in a web browser:

>     ```shell
>     [root@pappuses1 jpa]# odo url list
>     Found the following URLs for component mysboproj
>     NAME               STATE      URL                                           PORT     SECURE     KIND
>     mysboproj-9080     Pushed     http://mysboproj-9080.192.168.49.2.nip.io     9080     false      ingress
>     [root@pappuses1 jpa]#
>     ```

Reviewer 2: ```suggestion
    ```shell
   $ odo url list
    Found the following URLs for component mysboproj
    NAME               STATE      URL                                           PORT     SECURE     KIND
    mysboproj-9080     Pushed     http://mysboproj-9080.192.168.49.2.nip.io     9080     false      ingress
    ```
```

10. Use the URL to navigate to the `CreatePerson.xhtml` data entry page and enter requested data:
URL/CreatePerson.xhtml' and enter a user's name and age data using the form.

11. Click on the **Save** button when complete

Note that the entry of any data does not result in the data being displayed when you click on the "View Persons Record List" link.

### Creating a database to be used by the sample application

You can use the default configurations of the PostgreSQL Operator to start a Postgres database from it. But since our app uses few specific configuration values, lets make sure they are properly populated in the database service we start.

1. Store the YAML of the service in a file:
    ```shell

>     $ odo service create postgresql-operator.v0.1.1/Database --dry-run > db.yaml

Reviewer 2: ```suggestion
 odo service create postgresql-operator.v0.1.1/Database --dry-run > db.yaml
```

    ```

2. Modify and add following values under `metadata:` section in the `db.yaml` file:
    ```yaml
    name: sampledatabase
    annotations:
      service.binding/db_name: 'path={.spec.databaseName}'
      service.binding/db_password: 'path={.spec.databasePassword}'
      service.binding/db_user: 'path={.spec.databaseUser}'
    ```

    This configuration ensures that when a database service is started using this file, appropriate annotations are added to it. Annotations help the Service Binding Operator in injecting those values into the application. Hence, the above configuration will help Service Binding Operator inject the values for `databaseName`, `databasePassword` and `databaseUser` into the application.

3. Change the following values under `spec:` section of the YAML file:
    ```yaml
    databaseName: "sampledb"
    databasePassword: "samplepwd"
    databaseUser: "sampleuser"
    ```

> 4. Create the database from the YAML file:
>     ```shell
>     $ odo service create --from-file db.yaml
>     $ odo push
>     ```

Reviewer 2: ```suggestion
4. Create the database from the YAML file:
    ```shell
    odo service create --from-file db.yaml
    ```
    ```shell    
    odo push
    ```
```

   Operators available in the cluster
   NAME                                             CRDs
   postgresql-operator.v0.1.1                       Backup, Database

>     ```

Reviewer 2: Not sure why we need `odo catalog` set of commands here. We should use `odo service list` to list the Database service we created in above sectoin.

Reviewer 2: Since that's what point 2 in this section is about, IMO, we could skip the first part of `odo catalog list services` entirely.

Now, the only thing that remains is to connect the DB and the application. We will use odo to create a link to the Dev4Devs PostgreSQL Database Operator in order to access the database connection information.

1. Display the services available to odo: - You will see an entry for the PostgreSQL Database Operator displayed:
    ```shell
    $ odo catalog list services
    Operators available in the cluster
    NAME                                             CRDs
    postgresql-operator.v0.1.1                       Backup, Database

>     ```

Reviewer 2: This step is not required at this stage, IMO. @scottkurz WDYT?

2. List the service associated with the database created via the PostgreSQL Operator:
    ```shell
    $ odo service list
    NAME                       MANAGED BY ODO     STATE     AGE
    Database/sampledatabase   Yes (mysboproj)    Pushed    6m35s
    ```

3. Create a Service Binding Request between the application and the database using the Service Binding Operator service created in the previous step `odo link` command:
    ```shell

>     $ odo link Database/sampledatabase

Reviewer 2: ```suggestion
odo link Database/sampledatabase
```

    ```

4. Push this link to the cluster:
    ```shell

>     $ odo push

Reviewer 2: ```suggestion
odo push
```

---
title: Deploying a Java OpenLiberty application with PostgreSQL
sidebar_position: 1
---

This scenario illustrates deploying a Java application with odo and linking it to an in-cluster PostgreSQL service in the minikube environment.

There are two roles in this example:

> 1. Cluster Admin - Prepare the cluster by installing the required operators into the cluster.

Reviewer 2: ```suggestion
1. Cluster Admin - Prepare the cluster by installing the required operators on the cluster.
```

2. Application Developer - Imports a Java application, creates a Database instance, and connect the application to the Database instance.

## Cluster admin
---

> This section assumes that you have installed minikube and configured it. See the section on setting up a [Kubernetes](../getting-started/cluster-setup/kubernetes.md) cluster using minikube if you have not already configured it.

Reviewer 2: ```suggestion
This section assumes that you have installed [minikube and configured it](../getting-started/cluster-setup/kubernetes.md).
```

> The cluster admin must install two Operators into the cluster:

Reviewer 2: ```suggestion
The cluster admin must install two Operators on the cluster:
```

> The cluster admin must install two Operators into the cluster:
> 
> 1. PostgreSQL Operator
> 2. Service Binding Operator
> 
> An operator helps in deploying instances of a given service, for example PostgreSQL, MySQL, Redis.
> 
> Furthermore, these operators are "bind-able" as they expose information necessary for an application to connect to their instances.
> 
> See the [Operator installation guide](../getting-started/cluster-setup/kubernetes.md) to install and configure an operator in the Kubernetes cluster, if you have not already done so.
> 
> We will use [Dev4Devs PostgreSQL Operator](https://operatorhub.io/operator/postgresql-operator-dev4devs-com) found on the [OperatorHub](https://operatorhub.io) to demonstrate a sample use case.

Reviewer 2: ```suggestion
We will be using Operators in this guide. An Operator helps in deploying the instances of a given service, for example PostgreSQL, MySQL, Redis.

Furthermore, these Operators are "bind-able". Meaning, if they expose information necessary to connect to them, odo can help connect your component to their instances, 

See the [Operator installation guide](../getting-started/cluster-setup/kubernetes.md) to install and configure an Operator in the Kubernetes cluster, if you have not already done so.

The cluster admin must install two Operators into the cluster:

1. PostgreSQL Operator
2. Service Binding Operator

We will use [Dev4Devs PostgreSQL Operator](https://operatorhub.io/operator/postgresql-operator-dev4devs-com) found on the [OperatorHub](https://operatorhub.io) to demonstrate a sample use case.
```

## Application Developer
---

> This section assumes that you have installed `odo`. See the [Installation guide](../getting-started/installation.md) if you have not already installed it.

Reviewer 2: ```suggestion
This section assumes that you have [installed `odo`](../getting-started/installation.md).
```

Since the PostgreSQL Operator installed in above step is available only in `my-postgresql-operator-dev4devs-com` namespace, ensure that `odo` uses this namespace to perform any tasks:

```shell
odo project set my-postgresql-operator-dev4devs-com
```
### Importing the demo Java MicroService JPA application

In this example we will use odo to manage a sample [Java MicroServices JPA application](https://github.com/OpenLiberty/application-stack-samples.git).

1. Clone the sample application to your system:
    ```shell
    git clone https://github.com/OpenLiberty/application-stack-samples.git
    ```

2. Go to the sample JPA app directory:
    ```shell
    cd ./application-stack-samples/jpa
    ```

> 3. Initialize the project:

Reviewer 2: ```suggestion
3. Initialize the component:
```

Author: Wouldn't it be `Initialize the application`?

    ```shell
    odo create java-openliberty mysboproj
    ```

>    `java-openliberty` is the type of your application and `mysboproject` is the name of your application.

Author: Why not `application`? I thought we were trying to minimize the use of terms such as component and devfile.

>    `java-openliberty` is the type of your application and `mysboproject` is the name of your application.

Author: Decided to go with `application`.

4. Push the application to the cluster:
    ```shell

>     odo push --show-logs

Author: Good catch!

    ```

> 5. The application is now deployed to the cluster - you can view the status of the cluster, and the application test results by streaming the cluster logs into the terminal.

Reviewer 2: `odo log` doesn't show the status of the cluster. And streaming the logs doesn't happen unless one uses `--follow` flag with `odo log`. And it doesn't stream "cluster logs". It streams the logs of the component that we pushed to the cluster in the previous step.

    [INFO] [err]    ... 86 more
    [ERROR] Tests run: 2, Failures: 1, Errors: 1, Skipped: 0, Time elapsed: 0.706 s <<< FAILURE! - in org.example.app.it.DatabaseIT
    [ERROR] testGetAllPeople  Time elapsed: 0.33 s  <<< FAILURE!
    org.opentest4j.AssertionFailedError: Expected at least 2 people to be registered, but there were only: [] ==> expected: <true> but was: <false>
    at org.example.app.it.DatabaseIT.testGetAllPeople(DatabaseIT.java:57)
    
    [ERROR] testGetPerson  Time elapsed: 0.047 s  <<< ERROR!
    java.lang.NullPointerException
    at org.example.app.it.DatabaseIT.testGetPerson(DatabaseIT.java:41)
    
    [INFO]
    [INFO] Results:
    [INFO]
    [ERROR] Failures:
    [ERROR]   DatabaseIT.testGetAllPeople:57 Expected at least 2 people to be registered, but there were only: [] ==> expected: <true> but was: <false>
    [ERROR] Errors:
    [ERROR]   DatabaseIT.testGetPerson:41 NullPointer
    [INFO]
    [ERROR] Tests run: 2, Failures: 1, Errors: 1, Skipped: 0
    [INFO]
    [ERROR] Integration tests failed: There are test failures.
    ```
   Note: This error will be fixed at a later stage in the tutorial when we connect a database instance to this application.

> 6. You can also create an ingress URL with `odo` to access the application:

Reviewer 2: ```suggestion
6. You can also create an URL with `odo` to access the application:
```
`ingress` is a k8s lingo. We should avoid using it in odo docs.

    ```shell
    odo url create --host $(minikube ip).nip.io
    ```

7. Push the URL to activate it:
    ```shell

>     odo push --show-logs

Reviewer 2: ```suggestion
    odo push --show-log
```

    ```

8. Display the created URL:
    ```shell
    odo url list
    ```

    You will see a fully formed URL that can be used in a web browser:
    ```shell
    $ odo url list
    Found the following URLs for component mysboproj
    NAME               STATE      URL                                           PORT     SECURE     KIND
    mysboproj-9080     Pushed     http://mysboproj-9080.192.168.49.2.nip.io     9080     false      ingress
    ```

> 10. Use the URL to navigate to the `CreatePerson.xhtml` data entry page and enter requested data:

Reviewer 2: ```suggestion
10. Use the URL to navigate to the `CreatePerson.xhtml` data entry page to use the application:
```

> URL/CreatePerson.xhtml and enter a user's name and age data using the form.

Reviewer 2: ```suggestion
in case of this tutorial, we will access `http://mysboproj-9080.192.168.49.2.nip.io/CreatePerson.xhtml`. Note that the URL could be different for you. Now, enter a name and age data using the form.
```

11. Click on the **Save** button when complete

> Note that the entry of any data does not result in the data being displayed when you click on the "View Persons Record List" link.

Reviewer 2: Why? Because we haven't yet connected a database? If yes, please elaborate the reason and mention that we will be doing that in the next section. Just like we've mentioned similar message earlier in the doc.
